src/service/vsync-impl.js

```diff
@@ -224,13 +242,28 @@ export class Vsync {
   }
 
   /**
-   * @param {!Node=} unusedOptContextNode
+   * @param {!Node=} opt_contextNode
    * @return {boolean}
    * @private
    */
-  canAnimate_(unusedOptContextNode) {
-    // TODO(dvoytenko, #3742): Use opt_node -> ampdoc.
-    return this.viewer_.isVisible();
+  canAnimate_(opt_contextNode) {
+    // Window level: animations allowed only when global window is visible.
+    if (this.docState_.isHidden()) {
+      return false;
+    }
+
+    // Single doc: animations allowed when single doc is visible.
```

Comment:
This will be fine. When doc is available, the `viewer` is guaranteed to be available. The risk condition is ok as well - it will simply yield the same instance in this case. The only place where viewer might not be available is the `constructor` above.